jetty-server/src/main/java/org/eclipse/jetty/server/AsyncContentProducer.java

```diff
@@ -0,0 +1,352 @@
+    }
+
+    @Override
+    public HttpInput.Content nextContent()
+    {
+        HttpInput.Content content = nextTransformedContent();
+        if (LOG.isDebugEnabled())
+            LOG.debug("nextContent = {}", content);
+        if (content != null)
+            _httpChannel.getState().onReadIdle();
+        return content;
+    }
+
+    @Override
+    public void reclaim(HttpInput.Content content)
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("reclaim {} [t={}]", content, _transformedContent);
+        if (_transformedContent == content)
+        {
+            content.succeeded();
+            if (_transformedContent == _rawContent)
+                _rawContent = null;
+            _transformedContent = null;
+        }
+    }
+
+    @Override
+    public boolean isReady()
+    {
+        HttpInput.Content content = nextTransformedContent();
+        if (content == null)
+        {
+            _httpChannel.getState().onReadUnready();
+            if (_httpChannel.needContent())
+            {
+                content = nextTransformedContent();
+                if (LOG.isDebugEnabled())
+                    LOG.debug("isReady got transformed content after needContent retry {}", content);
+                if (content != null)
+                    _httpChannel.getState().onContentAdded();
+            }
+            else
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("isReady has no transformed content after needContent");
+            }
+        }
+        else
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("isReady got transformed content {}", content);
+            _httpChannel.getState().onContentAdded();
+        }
+        boolean ready = content != null;
+        if (LOG.isDebugEnabled())
+            LOG.debug("isReady = {}", ready);
+        return ready;
+    }
+
+    private HttpInput.Content nextTransformedContent()
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("nextTransformedContent [r={},t={}]", _rawContent, _transformedContent);
+        if (_rawContent == null)
+        {
+            _rawContent = produceRawContent();
+            if (_rawContent == null)
+                return null;
+        }
+
+        if (_transformedContent != null && _transformedContent.isEmpty())
+        {
+            if (_transformedContent != _rawContent)
+                _transformedContent.succeeded();
+            if (LOG.isDebugEnabled())
+                LOG.debug("nulling depleted transformed content");
+            _transformedContent = null;
+        }
+
+        while (_transformedContent == null)
+        {
+            if (_rawContent.isSpecial())
+            {
+                _error = _rawContent.getError() != null;
+                if (LOG.isDebugEnabled())
+                    LOG.debug("raw content is special (with error = {}), returning it", _error);
+                return _rawContent;
```

Comment:
Doesn't special content, or at least EOF, need to be passed to the interceptors?
Remember that we can EOF on non special content now, so interceptors might sometimes see EOF.... so they should always see it.
It may be that an interceptor is buffering content and can produce lots of content (over and over) from an empty EOF raw content.

Reply:
Our one interceptor does not need this, but I think it is needed in general.  So let's not make it a show stopper, but can you add a TODO and open a new issue.

Reply:
I added a todo about this, let's move that out of this PR.